packages/material-ui/src/OverridableComponent.d.ts

```diff
@@ -0,0 +1,39 @@
+import * as React from 'react';
+import { ConsistentWith, PropsOf, Omit } from '.';
+import { StyledComponentProps } from './styles';
+
+export interface OverridableComponent<M extends OverridableTypeMap> {
+  <C extends AnyReactType>(props: { component: C } & OverrideProps<M, C>): JSX.Element;
+  (props: DefaultProps<M>): JSX.Element;
+}
+
+// prettier-ignore
+export type OverrideProps<
+  M extends OverridableTypeMap,
+  C extends AnyReactType
+> = (
+  & BaseProps<M>
+  & Omit<PropsOf<C>, keyof StyledComponentProps<M['classKey']>>
+);
+
+// prettier-ignore
+export type DefaultProps<M extends OverridableTypeMap> =
+  & BaseProps<M>
+  & Omit<PropsOf<M['defaultComponent']>, keyof BaseProps<M>>;
+
+// prettier-ignore
+export type BaseProps<M extends OverridableTypeMap> =
+  & M['props']
+  & StyledComponentProps<M['classKey']>;
+
+export interface OverridableTypeMap {
```

Comment:
I really dig that approach. It's basically named parameters for generic arguments. Helps so much when understanding typings.